### PR TITLE
feat: First implementation of Sqlite DB service

### PR DESCRIPTION
**Pull request by nicolatimeus**
Signed-off-by: Nicola Timeus <nicola.timeus@eurotech.com>

> **Note**: We are using the Conventional Commits convention for our pull request titles. Please take a look at the [PR title format document](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#submitting-the-changes) for the supported [types](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#type) and [scopes](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#scope).

Adds a first implementation of a `BaseDbService` based on SQLite.

Adds the following dependency:

`maven/mavencentral/org.xerial/sqlite-jdbc/3.39.3.0, Apache-2.0 AND BSD-2-Clause AND ISC AND Artistic-2.0, approved, #3763`

Implemented features:
* In memory and persisted mode
* Support for WAL and rollback journal journaling modes
* Support for periodic defrag (VACUUM) and periodic WAL checkpoint.

Known issues:
* The current implementation of `DataService` will not work, due to incompatible queries
* The current implementation of the generic Database Wire Components will work only partially due to incompatible queries, (WireRecords can be stored in a SQLite database, but record removal will not work, causing the file size to grow indefinitely)
* Sporadic exception are visible in the logs reporting that a rollback operation failed. It seems that the connection pool implementation provided by the SQLite Java wrapper used here always performs a rollback when a pooled connection is returned, if a transaction is not open this will fail
* Exceptions in in-memory mode if concurrent queries are performed on the same table.

Tests are still missing.
Currently the SQLite provider bundle is not added to the distribution, but it is included in a deployment package instead.

### CODE REVIEW

**Comment by nicolatimeus** [comment]
Should be ready for review, once done we should squash any new commits and use Rebase and Merge so that the first commit that tracks the new dependency is preserved.

**Review by salvatore-coppola — APPROVED** [review]

File: kura/org.eclipse.kura.db.sqlite.provider/OSGI-INF/metatype/org.eclipse.kura.internal.db.sqlite.provider.SqliteDbServiceImpl.xml [hunk 1]
```diff
@@ -0,0 +1,91 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+    Copyright (c) 2022 Eurotech and/or its affiliates and others
+  
+    This program and the accompanying materials are made
+    available under the terms of the Eclipse Public License 2.0
+    which is available at https://www.eclipse.org/legal/epl-2.0/
+ 
+	SPDX-License-Identifier: EPL-2.0
```
Comment (MMaiero) [inline kura/org.eclipse.kura.db.sqlite.provider/OSGI-INF/metatype/org.eclipse.kura.internal.db.sqlite.provider.SqliteDbServiceImpl.xml]: The formatting seems odd

File: kura/org.eclipse.kura.db.sqlite.provider/OSGI-INF/metatype/org.eclipse.kura.internal.db.sqlite.provider.SqliteDbServiceImpl.xml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+         name="SqliteDbService" 
+         description="Sqlite based database service.">
+
+        <AD id="db.mode"
+            name="Database Mode"
+            type="String"
+            cardinality="0" 
+            required="true"
+            default="IN_MEMORY"
+            description="Defines the database mode.">
```
Comment (MMaiero) [inline kura/org.eclipse.kura.db.sqlite.provider/OSGI-INF/metatype/org.eclipse.kura.internal.db.sqlite.provider.SqliteDbServiceImpl.xml]: Do we want to expand the description to make it more comprehensive?

File: kura/org.eclipse.kura.db.sqlite.provider/OSGI-INF/metatype/org.eclipse.kura.internal.db.sqlite.provider.SqliteDbServiceImpl.xml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            <Option label="Persisted" value="PERSISTED"/>
+         </AD>
+         
+         <AD id="db.path"
+            name="Persisted Database Path"
+            type="String"
+            cardinality="0" 
+            required="true"
+            default="mydb"
+            description="Defines the database path. This parameter is only relevant for persisted databases.">
```
Comment (MMaiero) [inline kura/org.eclipse.kura.db.sqlite.provider/OSGI-INF/metatype/org.eclipse.kura.internal.db.sqlite.provider.SqliteDbServiceImpl.xml]: Do we want to state that we are expecting a full path, not relative?

File: kura/org.eclipse.kura.db.sqlite.provider/OSGI-INF/metatype/org.eclipse.kura.internal.db.sqlite.provider.SqliteDbServiceImpl.xml [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+         <AD id="db.wal.checkpoint.interval.seconds"
+            name="WAL Checkpoint Interval (Seconds)"
+            type="Long"
+            cardinality="0" 
+            required="true"
+            default="600"
+            min="0"
+            description="SqliteDbService instances support running periodic periodic WAL checkpoints (https://www.sqlite.org/pragma.html#pragma_wal_checkpoint). Checkpoints will be performed in TRUNCATE mode. This parameter specifies the interval in seconds beetween two consecutive checkpoints, set to zero to disable. This parameter is only relevant for persisted databases in WAL Journal Mode."/>
+
+         <AD id="db.connection.pool.max.size"
```
Comment (MMaiero) [inline kura/org.eclipse.kura.db.sqlite.provider/OSGI-INF/metatype/org.eclipse.kura.internal.db.sqlite.provider.SqliteDbServiceImpl.xml:67]: Would it make sense to move this and the following one up in the metatype as it affects both in memory and persistent databases? What is the frequency of usage of such properties for a normal setup?

File: kura/org.eclipse.kura.db.sqlite.provider/OSGI-INF/org.eclipse.kura.internal.db.sqlite.provider.SqliteDbServiceImpl.xml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+   Copyright (c) 2022 Eurotech and/or its affiliates and others
+  
+   This program and the accompanying materials are made
+   available under the terms of the Eclipse Public License 2.0
+   which is available at https://www.eclipse.org/legal/epl-2.0/
+ 
+	SPDX-License-Identifier: EPL-2.0
```
Comment (MMaiero) [inline kura/org.eclipse.kura.db.sqlite.provider/OSGI-INF/org.eclipse.kura.internal.db.sqlite.provider.SqliteDbServiceImpl.xml]: Formatting seems odd

File: kura/org.eclipse.kura.db.sqlite.provider/OSGI-INF/org.eclipse.kura.internal.db.sqlite.provider.SqliteDebugShell.xml [hunk 6]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+   Copyright (c) 2022 Eurotech and/or its affiliates and others
+  
+   This program and the accompanying materials are made
+   available under the terms of the Eclipse Public License 2.0
+   which is available at https://www.eclipse.org/legal/epl-2.0/
+ 
+	SPDX-License-Identifier: EPL-2.0
```
Comment (MMaiero) [inline kura/org.eclipse.kura.db.sqlite.provider/OSGI-INF/org.eclipse.kura.internal.db.sqlite.provider.SqliteDebugShell.xml]: Also here

File: kura/org.eclipse.kura.db.sqlite.provider/OSGI-INF/metatype/org.eclipse.kura.internal.db.sqlite.provider.SqliteDbServiceImpl.xml (hunk 4 above)
Comment (nicolatimeus) [reply kura/org.eclipse.kura.db.sqlite.provider/OSGI-INF/metatype/org.eclipse.kura.internal.db.sqlite.provider.SqliteDbServiceImpl.xml:67]: Maybe it is fine as is, the user probably will not change frequently the connection pool size, and the last parameter is just for debug purposes

**Comment by nicolatimeus** [comment]
The other comments should have been addressed. If it is ok to merge, please use the rebase and merge feature and not squash the commits.

File: kura/features/org.eclipse.kura.db.sqlite.provider/pom.xml [hunk 7]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (MMaiero) [inline kura/features/org.eclipse.kura.db.sqlite.provider/pom.xml:1]: Copyright header missing

File: kura/org.eclipse.kura.db.sqlite.provider/OSGI-INF/metatype/org.eclipse.kura.internal.db.sqlite.provider.SqliteDbServiceImpl.xml (hunk 4 above)
Comment (MMaiero) [reply kura/org.eclipse.kura.db.sqlite.provider/OSGI-INF/metatype/org.eclipse.kura.internal.db.sqlite.provider.SqliteDbServiceImpl.xml:67]: Agreed

**Review by MMaiero — APPROVED** [review]

**Pull request merged**